source/_posts/2016-11-08-Vapor-mysql.md

```diff
@@ -0,0 +1,195 @@
+
+### 2. Setting up the DropLet to support the provider
+
+Now open your main.swift file and add the following lines of code:
+
+```swift
+...
+//Importing our new MySQL provider
+import VaporMySQL
+
+let drop = Droplet()
+
+// Add providers. This tells Vapor that we are using the VaporMySQL provider, this will bind the data to the database and the models automaticlly down the line
+try drop.addProvider(VaporMySQL.Provider.self)
+
+//Making sure that Vapor runs our migrations / preperations for our model(s) 
+drop.preparations.append(Car.self)
+
+....
+
+drop.run()
+```
+Now lets jump into our `Car.swift` file (and down in the extension of the model) add some actual Preparation code. 
+
+```swift
+extension Car: Preparation {
+    static func prepare(_ database: Database) throws {
+        
+        //Adding our actual migration table and attributes. We are first defining the name of the database table and afterwards what attributes the table should have.
+        try database.create("cars") { cars in
+            cars.id()
+            cars.data("name")
+            cars.data("color")
+            cars.int("miles_driven")
+        }
+    }
+    
+    //This makes sure it gets deleted when reverting the projects database
+    static func revert(_ database: Database) throws {
+        try database.delete("cars")
+    }
+}
+```
+
+If you now Run the project then it will compile correct but crash right after it has started. This is due to we are missing a MySQL config file. 
+
+So now open terminal again and run:
+
+`touch Config/mysql.json`
+
+Open the newly added file and add the following: (Of course depending on your local MySQL settings)
+
+```bash
+{
+    "host": "127.0.0.1",
+    "user": "root",
+    "password": "",
+    "database": "drive"
+}
+```
+
+If you try to Run your Xcode project now and get the following:
+
+```bash
+Failed to start preparation
+Can\'t connect to MySQL server on \'127.0.0.1\' (61)
+```
+
+Then run `mysql.server start` in your terminal, this makes sure that you have MySQL running on your local machine. You will see a success message when it's up running. (You can't continue the tutorial before you have this setup)
+
+When you have a MySQL database setup on your local machine, then try to Run the Xcode project. You should then see the following:
+
+```bash
+Preparing Car
+Prepared Car
+Database prepared
+Server 'default' starting at 0.0.0.0:8080
+```
+
+The consol are here telling you that the model Car is getting setup in the database as a table. You will only get this message(s) the first time you Run the project where you don't have the tables setup in the database. 
+
+And boom, we are up and running. Now check your database interface if you are using any (I'm using [SequelPro](https://www.sequelpro.com/)), you should now be able to see our Car model as a database table (cars). You will also see a table called 'fluent', this is Vapors way of keeping track of what models should be linked to what tables.
```

Comment:
It's Vapors way of keeping track of which tables have been created, not which models should be linked to what tables.